src/main/java/pl/edu/pw/mini/ingreedio/api/controller/ProductController.java

```diff
@@ -20,55 +20,49 @@
 import org.springframework.web.bind.annotation.RequestParam;
 import org.springframework.web.bind.annotation.ResponseBody;
 import org.springframework.web.bind.annotation.RestController;
-import pl.edu.pw.mini.ingreedio.api.criteria.ProductFilterCriteria;
 import pl.edu.pw.mini.ingreedio.api.dto.FullProductDto;
-import pl.edu.pw.mini.ingreedio.api.dto.ProductDto;
 import pl.edu.pw.mini.ingreedio.api.dto.ProductListResponseDto;
-import pl.edu.pw.mini.ingreedio.api.mapper.ProductDtoMapper;
 import pl.edu.pw.mini.ingreedio.api.model.Product;
+import pl.edu.pw.mini.ingreedio.api.service.PaginationService;
 import pl.edu.pw.mini.ingreedio.api.service.ProductService;
+import pl.edu.pw.mini.ingreedio.api.service.ProductsCriteriaService;
 
 @RestController
 @RequestMapping("/api/products")
 @RequiredArgsConstructor
 @Tag(name = "Products" /*, description = "..."*/)
 public class ProductController {
     private final ProductService productService;
-    private final ProductDtoMapper productDtoMapper;
-    private final int pageSize = 10;
+    private final PaginationService paginationService;
+    private final ProductsCriteriaService productsCriteriaService;
 
     @Operation(summary = "Get matching products",
         description = "Get matching products",
         security = {@SecurityRequirement(name = "Bearer Authentication")})
     @GetMapping
     public ResponseEntity<ProductListResponseDto> getProducts(
-        int pageNumber,
-        @RequestParam Optional<String> name,
-        @RequestParam Optional<String> provider,
-        @RequestParam Optional<String> brand,
-        @RequestParam Optional<Integer> volumeFrom,
-        @RequestParam Optional<Integer> volumeTo,
-        @RequestParam Optional<String[]> ingredients) {
-        PageRequest pageRequest = PageRequest.of(pageNumber, pageSize);
+        @RequestParam("page-number") Optional<Integer> pageNumber,
+        @RequestParam("ingredients-exclude") Optional<List<Long>> ingredientsToExclude,
+        @RequestParam("ingredients-include") Optional<List<Long>> ingredientsToInclude,
+        @RequestParam("min-rating") Optional<Integer> minRating,
+        @RequestParam("phrase") Optional<String> phrase,
+        @RequestParam("sort-by") Optional<List<String>> sortBy,
+        @RequestParam("liked") Optional<Boolean> liked) {
```

Comment:
In connection to one of my previous comments, can't we use here `Optional<Set<...>>`? Or is that prohibited by spring? If so, i think the conversion from `List` to `Set` should be done here, in the controller.

Reply:
I've tested Set and for some reason the fields where null, I will try to reproduce it

Reply:
> I've tested Set and for some reason the fields where null, I will try to reproduce it.

😭 Spring = flop

Reply:
> > I've tested Set and for some reason the fields where null, I will try to reproduce it.
> 
> 😭 Spring = flop

Actually it was mine flop this time, couldn't reproduce the error, so the `Set` is used from now